File: S20184060/HW11/HW11.py
import numpy as np
import scipy.constants as const
import matplotlib.pyplot as plt

Lz = 5e-9
mxx = 0.19*const.m_e
myy = 0.19*const.m_e
mzz = 0.91*const.m_e
T = 300.

n = np.arange(1,101)
Ez = (const.hbar*np.pi*n)**2/(2.*mzz*Lz*Lz)
Ef = np.linspace(-0.3,0.1,100).reshape((100,1))
ez = (const.e*Ef-Ez)/(const.k*T)

N_list = np.sum(1./np.pi*np.sqrt(mxx*myy)/const.hbar**2*const.k*T*np.log(1+np.exp(ez)),axis=1)

plt.figure(figsize=(10,10))
plt.plot(Ef.flatten(),N_list,'r-')

plt.xlabel('$E_F$',fontsize=18)
plt.ylabel('$n_{tot}(E_F)$',fontsize=18)
plt.yscale('log')

plt.show()

File: s20194005/HW11/HW11.m
hbar = 1.054571e-34;
q = 1.602192e-19;
m0 = 9.109534e-31;
T = 300;
k_B = 1.380662e-23;
Lx = 100e-9; 
Ly = 100e-9; 
Lz = 5e-9;
nmax = 50;
mxx=0.19;
myy=0.19;
mzz=0.91;
coef=2*Lx*Ly/(2*pi)*sqrt(mxx*myy)*m0/(hbar^2)*(k_B*T);
totalNumber=zeros(41,1);
for i=1:41
    Ef=0.01*i-0.31;
    for n=1:nmax
       Ez = (hbar^2)/(2*mzz*m0)*(pi*n/Lz)^2; 
       subbandNumber = coef*log(1+exp((-Ez+q*Ef)/(k_B*T)));
       totalNumber(i,1)= totalNumber(i,1)+ subbandNumber;
    end
end

semilogy(-0.3:0.01:0.1,totalNumber/(Lx*Ly)*1e-4, '-o');
xlabel('Fermi energy (eV)')
ylabel('Electron density (cm^-2)')
